File: packages/core/src/logger.ts
/**
 * Structured logging utilities for Plinto
 * Replaces console.log with structured, level-based logging
 */

export enum LogLevel {
  DEBUG = 'debug',
  INFO = 'info',
  WARN = 'warn',
  ERROR = 'error',
}

export interface LogContext {
  [key: string]: any;
}

export interface LogEntry {
  timestamp: string;
  level: LogLevel;
  message: string;
  context?: LogContext;
  error?: Error;
}

export class Logger {
  private context: LogContext = {};
  private minLevel: LogLevel;

  constructor(minLevel: LogLevel = LogLevel.INFO) {
    this.minLevel = minLevel;
  }

  /**
   * Set persistent context for all log entries
   */
  setContext(context: LogContext): void {
    this.context = { ...this.context, ...context };
  }

  /**
   * Clear persistent context
   */
  clearContext(): void {
    this.context = {};
  }

  /**
   * Check if log level should be output
   */
  private shouldLog(level: LogLevel): boolean {
    const levels = [LogLevel.DEBUG, LogLevel.INFO, LogLevel.WARN, LogLevel.ERROR];
    return levels.indexOf(level) >= levels.indexOf(this.minLevel);
  }

  /**
   * Format log entry as JSON
   */
  private format(entry: LogEntry): string {
    return JSON.stringify(entry);
  }

  /**
   * Output log entry
   */
  private output(level: LogLevel, message: string, context?: LogContext, error?: Error): void {
    if (!this.shouldLog(level)) return;

    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      message,
      context: { ...this.context, ...context },
      error: error ? {
        name: error.name,
        message: error.message,
        stack: error.stack,
      } as any : undefined,
    };

    const formatted = this.format(entry);

    // Output to appropriate stream
    if (level === LogLevel.ERROR || level === LogLevel.WARN) {
      console.error(formatted);
    } else {
      console.log(formatted);
    }
  }

  /**
   * Log debug message
   */
  debug(message: string, context?: LogContext): void {
    this.output(LogLevel.DEBUG, message, context);
  }

  /**
   * Log info message
   */
  info(message: string, context?: LogContext): void {
    this.output(LogLevel.INFO, message, context);
  }

  /**
   * Log warning message
   */
  warn(message: string, context?: LogContext, error?: Error): void {
    this.output(LogLevel.WARN, message, context, error);
  }

  /**
   * Log error message
   */
  error(message: string, context?: LogContext, error?: Error): void {
    this.output(LogLevel.ERROR, message, context, error);
  }

  /**
   * Create child logger with additional context
   */
  child(context: LogContext): Logger {
    const child = new Logger(this.minLevel);
    child.setContext({ ...this.context, ...context });
    return child;
  }
}

/**
 * Default logger instance
 */
export const logger = new Logger(
  process.env.NODE_ENV === 'production' ? LogLevel.INFO : LogLevel.DEBUG
);

/**
 * Create a logger for a specific module
 */
export function createLogger(module: string): Logger {
  return logger.child({ module });
}

File: apps/api/app/core/logger.py
"""
Structured logging configuration for Janua API
Replaces print() statements with structured JSON logging
"""

import logging
import sys
from datetime import datetime
from typing import Any, Dict, Optional
from pythonjsonlogger import jsonlogger


class CustomJsonFormatter(jsonlogger.JsonFormatter):
    """
    Custom JSON formatter that adds timestamp and formats log records
    """

    def add_fields(
        self, log_record: Dict[str, Any], record: logging.LogRecord, message_dict: Dict[str, Any]
    ) -> None:
        super().add_fields(log_record, record, message_dict)

        # Add timestamp
        log_record["timestamp"] = datetime.utcnow().isoformat() + "Z"

        # Add log level
        log_record["level"] = record.levelname

        # Add module info
        log_record["module"] = record.module
        log_record["function"] = record.funcName

        # Add any extra fields
        if hasattr(record, "user_id"):
            log_record["user_id"] = record.user_id
        if hasattr(record, "organization_id"):
            log_record["organization_id"] = record.organization_id
        if hasattr(record, "request_id"):
            log_record["request_id"] = record.request_id


def setup_logging(level: str = "INFO", json_output: bool = True) -> None:
    """
    Configure structured logging for the application

    Args:
        level: Logging level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        json_output: If True, output JSON format; otherwise use standard format
    """
    root_logger = logging.getLogger()
    root_logger.setLevel(getattr(logging, level.upper()))

    # Remove existing handlers
    root_logger.handlers.clear()

    # Create console handler
    handler = logging.StreamHandler(sys.stdout)

    if json_output:
        # JSON formatter for production
        formatter = CustomJsonFormatter(
            "%(timestamp)s %(level)s %(module)s %(funcName)s %(message)s"
        )
    else:
        # Standard formatter for development
        formatter = logging.Formatter(
            "%(asctime)s - %(name)s - %(levelname)s - %(message)s", datefmt="%Y-%m-%d %H:%M:%S"
        )

    handler.setFormatter(formatter)
    root_logger.addHandler(handler)


def get_logger(name: str) -> logging.Logger:
    """
    Get a logger instance for a specific module

    Args:
        name: Logger name (typically __name__)

    Returns:
        Configured logger instance
    """
    return logging.getLogger(name)


class LoggerAdapter(logging.LoggerAdapter):
    """
    Logger adapter that adds contextual information to all log records
    """

    def process(self, msg: str, kwargs: Dict[str, Any]) -> tuple:
        # Add extra context to all log records
        if "extra" not in kwargs:
            kwargs["extra"] = {}

        # Merge adapter's extra fields
        for key, value in self.extra.items():
            kwargs["extra"][key] = value

        return msg, kwargs


def get_context_logger(
    name: str,
    user_id: Optional[str] = None,
    organization_id: Optional[str] = None,
    request_id: Optional[str] = None,
) -> LoggerAdapter:
    """
    Get a logger with contextual information

    Args:
        name: Logger name
        user_id: User ID to include in logs
        organization_id: Organization ID to include in logs
        request_id: Request ID to include in logs

    Returns:
        Logger adapter with context
    """
    logger = get_logger(name)
    extra = {}

    if user_id:
        extra["user_id"] = user_id
    if organization_id:
        extra["organization_id"] = organization_id
    if request_id:
        extra["request_id"] = request_id

    return LoggerAdapter(logger, extra)


# Example usage:
# from app.core.logger import get_logger, get_context_logger
#
# # Basic usage
# logger = get_logger(__name__)
# logger.info("User authenticated successfully", extra={"user_id": user_id})
#
# # Context logger (automatically adds context to all logs)
# logger = get_context_logger(__name__, user_id=user.id, request_id=request_id)
# logger.info("Password reset initiated")  # Automatically includes user_id and request_id
